### DISCUSSION TITLE
no _boot.py after compiling

### DISCUSSION THREAD

**Discussion by martijnzm** [micropython/micropython#13337] [ESP8266]
In order to freeze additional modules in the the firmware I am compiling µPython from source following the instruction on [https://github.com/micropython/micropython/tree/master/ports/esp8266|github]. After some effort (main lesson: don't use the Espressif SDK) I have created and flashed firmware. But after boot, very little libraries are available, specifically mip and webrepl are not available, leaving no possibilities to add more libraries.

Info:
- Board: ESP8266 NodeMCU v3 with 4Mb flash
- pre built µPython images work, though I quickly run into memory problems using libraries like microdot
- Micropython checked out from github. I tried both master and v1.22.0
- esp_open_sdk version 2018-06-10 (why so old?)
- modules and mpy_cross successfully built
- compiled µPython using  `make -j BOARD=ESP8266_GENERIC`, with and without the variant of 1Mb flash and with or without FROZEN_MANIFEST to include microdot.
- `import webrepl` and `import mip` give ImportError (no module)
- `import os; os.listdir()' shows an empy file system, whereas a boot.py and webrepl_cfg.py should be present
- When using FROZEN_MANIFEST to freeze microdot, `import microdot` fails on `import asyncio`. So apparently microdot.mpy itself has successfully been included.
- After a soft reset 'ctrl-d', I get a message that _boot.py cannot be found.
- I tried to run the commands `from _boot.py import uos, from flashbdev import bdev, uos.VfsFat.mkfs(bdev)` manually but flashbdev is not available

My conclusion is that the "root filesystem" is missing or mostly empty. But it is strange to see that microdot.mpy can be accessed. So probably there is file system but many standard libraries are missing. I did 'make -C ports/esp8266 submodules' before compiling µPython so I don't understand why.

- Can someone point me to a the solution?
- How to access the filesystem containing all modules?

Thanks for the effort of replying. I hope I provided enough information.

**Comment by robert-hh** [comment]
The file _boot.py is in the modules directory. Therefore somewhere in your manifest.py chain there must be a statement `freeze("$(PORT_DIR)/modules")`.

**Comment by martijnzm** [comment]
Thanks!
The manifest file ESP8266_GENERIC\boards\manifest_2Mib.py contains a line `include("$(PORT_DIR)/boards/manifest.py")` which in turn contains a line `freeze("$(PORT_DIR)/modules")`
So the modules should be frozen as the line is in place. I added a print statement before the freeze line. After a make clean, I rebuilt the image, but the print statement did not show up. Apparently the manifest file is ignored.
My assumption was that specifying BOARD=ESP8266_GENERIC would result in calling of the correct manifest.py files.
Let's have look at the Makefile.

**Comment by robert-hh** [comment]
The default for FROZEN_MANIFEST for the ESP8266_GENERIC board is `FROZEN_MANIFEST ?= $(BOARD_DIR)/manifest_2MiB.py`. If you change FROZEN_MANIFEST to something else, you have to take care to include the previous chain of manifest files.
Using the board variants FLASH_1M or OTA will just use boards/manifest.py.

**Reply by robert-hh** [reply]
What is the output of `help ("modules")`? it shows all built-in and frozen modules.

**Comment by martijnzm** [comment]
```python
>>> help("modules")
__main__          cryptolib         lwip              select
_asyncio          deflate           machine           socket
_espnow           errno             math              ssl
_onewire          esp               microdot          struct
_webrepl          framebuf          micropython       sys
array             gc                network           time
binascii          hashlib           os                uctypes
btree             heapq             platform          websocket
builtins          io                random
collections       json              re
Plus any modules on the filesystem
``` 
So flashbdev is indeed missing, as is mip, but microdot is present.

**Reply by robert-hh** [reply]
How does not manifest.py file look like, which you use to include microdot?

**Reply by martijnzm** [reply]
So far it only contained two lines:
```
package("tinyweb")
module("microdot.py")'
``` 
But I did not include this file for all builds to avoid my own customisations to cause trouble. But since it is apparently used correctly I just added the line 'include("$(PORT_DIR)/boards/manifest.py")', and now the modules have been included:
```'Python
|========================================|
| Freezing modules                       |
|========================================|

MPY flashbdev.py
MPY apa102.py
MPY _boot.py
MPY inisetup.py
MPY port_diag.py
MPY espnow.py
MPY requests/__init__.py
MPY mip/__init__.py
MPY ntptime.py
MPY webrepl.py
MPY webrepl_setup.py
MPY urequests.py
MPY dht.py
MPY onewire.py
MPY ds18x20.py
MPY neopixel.py
```

**Reply by martijnzm** [reply]
Ah, now I see that boards/manifest.py is called when calling `make deploy`, not during build. Before I did not use make deploy since it exited with an error because my board does not support checking of checksum. I just ran esptool to upload firmware.bin, as described in README.md. But apparantly  Today I removed the --verify option in the Makefile, so deploy does work. And it finalises the image. Apparently the last steps of assembling the image are only done during deploying. Which is in my opinion not correct. It also looks like the  README is a bit misleading here. Or I missed a step somewhere...
At least it works now.

**Comment by robert-hh** [comment]
If you use a 4M module, you can instead include the one from BOARD_DIR:
`include("$(BOARD_DIR)/manifest_2MiB.py")`, which adds asyncio and umqtt and more.

**Reply by martijnzm** [reply]
How can I specify which manifest file is used during build?
I think I will retry from a completely clean tree to see when the manifest file is only called during deploy.

**Comment by robert-hh** [comment]
> Ah, now I see that boards/manifest.py is called when calling make deploy, not during build. 

That's not right. FROZEN_MANIFEST is set in the mpconfigboard.mk or Makefile, whether you call just `make` or `make deploy`. You can override the setting of FROZEN_MANIFEST in the make files by setting it on the command line calling make, e.g.:

`make -j FROZEN_MANIFEST=my_manifest_file ......`

You can also create a file called GNUmakefile, which sets your preferred overrides, like e.g.:
```
# GNUmakefile
BOARD ?= ESP8266_GENERIC
PORT ?= /dev/ttyUSB0
FROZEN_MANIFEST ?= my_manifest.py
include Makefile
```
Then, when you call `make .....`, the GNUmakefile is used.

**Reply by martijnzm** [reply]
Above I wrote that microdot is included in the build, but I found it isn't. Help("modules") does not show microdot and Microdot was working just because if was still present on the filesystem whereas I assumed the filesystem would be wiped out by flashing new firmware.
The frozen_mpy folder contains a microdot.mpy file, so the file is successfully compiled. Nevertheless, it does not show up in the firmware.bin. A `grep "microdot" fimware.bin` yields two references of the location of microdot.py. So I am wondering whether the firmware does contain the microdot mpy file, but I can't use it or that microdot is not included after all.
What could be wrong here?

**Comment by jimmo** [comment]
@martijnzm I'm having a bit of a hard time following the thread above.

In general anything that involves customising the build means you should make your own board definition.

If you copy the ESP8266_GENERIC directory (place the copy anywhere, doesn't have to be in the MicroPython repo), then you can edit the files as you reqire and build with

```
make BOARD_DIR=../../path/to/my/BOARD
```

When you copy the directory, you can remove `manifest_512kiB.py` and `_boot.py`, rename `manifest_2MiB.py` to just `manifest.py` and simplify `mpconfigboard.mk` (i.e. just take lines 2-26, and update line 10 to be `manifest.py`).

Then you can do whatever you like to your board's manifest.py
